### PR TITLE
doc: MetaM

### PR DESCRIPTION
**Pull request by EdAyers**

### CODE REVIEW

File: src/Lean/Meta/Basic.lean [hunk 1]
```diff
@@ -1297,6 +1334,7 @@ def isExprDefEq (t s : Expr) : MetaM Bool :=
     trace[Meta.isDefEq] "{t} =?= {s} ... {if b then "success" else "failure"}"
     return b
 
+/-- Determines whether two expressions are definitionally equal to each other. -/
```
Comment (EdAyers) [inline src/Lean/Meta/Basic.lean:1337]: Something I haven't managed to figure out yet is where the flags go to do unification and matching.

Comment (leodemoura) [reply src/Lean/Meta/Basic.lean:1337]: Which flags?

**Comment by leodemoura** [comment]
I am merging since the question above does not affect the PR.

File: src/Lean/Meta/Basic.lean (hunk 1 above)
Comment (EdAyers) [reply src/Lean/Meta/Basic.lean:1337]: I mean, how can I perform isDefEq without any mvar unification happening, or with only mvar assignment happening on rhs.

Comment (leodemoura) [reply src/Lean/Meta/Basic.lean:1337]: Metavariables and their context have a "depth". Given a metavariable `?m` and a `MetavarContext` `mctx`, `?m` is not assigned if `?m.depth != mctx.depth`. We also have a combinator `withNewMCtxDepth x` which will bump the `mctx` while executing `x`. So, `withNewMCtxDepth (<- isDefEq a b)` is `isDefEq` without any mvar assignment happening. For matching, we create the term inside the `withNewMCtxDepth`, here is an example from `simp`: https://github.com/leanprover/lean4/blob/master/src/Lean/Meta/Tactic/Simp/Rewrite.lean#L100-L106

**Pull request merged**